tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/iterator/BindLeftJoinIteration.java

```diff
@@ -0,0 +1,99 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Eclipse RDF4J contributors.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *******************************************************************************/
+package org.eclipse.rdf4j.federated.evaluation.iterator;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Set;
+
+import org.eclipse.rdf4j.common.iteration.CloseableIteration;
+import org.eclipse.rdf4j.common.iteration.LookAheadIteration;
+import org.eclipse.rdf4j.query.Binding;
+import org.eclipse.rdf4j.query.BindingSet;
+import org.eclipse.rdf4j.query.QueryEvaluationException;
+import org.eclipse.rdf4j.query.algebra.evaluation.QueryBindingSet;
+
+/**
+ * A {@link LookAheadIteration} for processing bind left join results.
+ *
+ * Algorithm:
+ *
+ * <ul>
+ * <li>execute left bind join using regular bound join query</li>
+ * <li>process result iteration similar to {@link BoundJoinVALUESConversionIteration}</li>
+ * <li>remember seen set of bindings (using index) and add original bindings to those, i.e. put to result return all
+ * non-seen bindings directly from the input</li>
+ *
+ *
+ * @author Andreas Schwarte
+ */
+public class BindLeftJoinIteration extends LookAheadIteration<BindingSet> {
+
+	protected final CloseableIteration<BindingSet> iter;
+	protected final List<BindingSet> bindings;
+
+	protected Set<Integer> seenBindingIndexes = new HashSet<>();
```

Comment:
Would the [IntHashSet](https://eclipse.dev/collections/javadoc/11.1.0/org/eclipse/collections/impl/set/mutable/primitive/IntHashSet.html) be appropriate here?

Reply:
Not sure what the impact of performance is, but it does the job. I guess this datastructure (just looking at its name is more optimized and hence more appropriate)

Reply:
It doesn't need to wrap the integer in an object (autoboxing?) so it uses a lot less memory and is overall a lot faster. Mostly the memory pressure that would be beneficial here.